src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -63,16 +62,161 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
+                return new BoundWildcardPattern(node, hasErrors: true);
+            }
+
+            // Note: we bind the pattern (conversion) to the first argument of `operator is`, not the type containing `operator is`.
+            // That is how we support the specification of pattern-matching external to a type (e.g. active patterns)
+            hasErrors = hasErrors || CheckValidPatternType(node.Type, operand, operandType, isOperator.Parameters[0].Type, false, diagnostics);
+            var patterns = ArrayBuilder<BoundPattern>.GetInstance(subPatternsCount);
+            for (int i = 0; i < subPatternsCount; i++)
+            {
+                if (subPatterns[i].NameColon != null)
+                {
+                    // PROTOTYPE(patterns): keyword syntax not supported in recursive patterns.
+                    diagnostics.Add(ErrorCode.ERR_FeatureIsUnimplemented, subPatterns[i].NameColon.Location, "keyword syntax in recursive patterns");
+                    hasErrors = true;
+                }
+
+                patterns.Add(BindPattern(subPatterns[i].Pattern, null, isOperator.Parameters[i + 1].Type, hasErrors, diagnostics));
+            }
+
+            return new BoundRecursivePattern(node, type, isOperator, patterns.ToImmutableAndFree(), hasErrors);
+        }
+
+        /// <summary>
+        /// Is a user-defined `operator is` well-formed? At the use site, we ignore those that are not.
+        /// </summary>
+        private bool WellFormedOperatorIs(MethodSymbol candidate, CSharpSyntaxNode node, DiagnosticBag diagnostics)
+        {
+            // requires at least one parameter
+            if (candidate.ParameterCount == 0)
+            {
+                return false;
+            }
+
+            // must be static.
+            // PROTOTYPE(patterns): we don't support value-capturing matchers (e.g. regular expressions) yet.
+            if (!candidate.IsStatic)
+            {
+                return false;
+            }
+
+            // the first parameter must be a value. The remaining parameters must be out.
+            foreach (var parameter in candidate.Parameters)
+            {
+                if (parameter.RefKind != ((parameter.Ordinal == 0) ? RefKind.None : RefKind.Out))
+                {
+                    return false;
+                }
+            }
+
+            // must return void or bool
+            switch (candidate.ReturnType.SpecialType)
+            {
+                case SpecialType.System_Void:
+                case SpecialType.System_Boolean:
+                    break;
+                default:
+                    return false;
+            }
+
+            // must not be generic
+            if (candidate.Arity != 0)
+            {
+                return false;
+            }
+
+            // it should be accessible
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            bool isAccessible = this.IsAccessible(candidate, ref useSiteDiagnostics);
```

Comment:
I'm not making an assumption, I'm checking whether or not it is accessible. I'm not intending to enforce the language constraints on the declarations at this point. I'm trying to select the most reasonable behavior at the use site even when declarations don't satisfy all invariants. In error recovery situations (e.g. when `operator is` was declared `internal` in source) I'd rather it not trigger another error here. In other words, this is the intended behavior.